File: app/main/geometry.ts
// Pure geometry helpers for window persistence. No Electron imports here so this
// module is unit-testable under plain Node (vitest).

export interface Geometry {
  x: number
  y: number
  width: number
  height: number
}

export interface Bounds {
  x: number
  y: number
  width: number
  height: number
}

export const MIN_WIDTH = 360
export const MIN_HEIGHT = 360

// A window counts as on-screen if it overlaps the monitor by at least 50px on
// both axes — enough to be grabbable. Mirrors the old Tauri windowState logic.
export function isOnScreen(g: Geometry, m: Bounds): boolean {
  const gRight = g.x + g.width
  const gBottom = g.y + g.height
  const mRight = m.x + m.width
  const mBottom = m.y + m.height

  const overlapX = Math.max(0, Math.min(gRight, mRight) - Math.max(g.x, m.x))
  const overlapY = Math.max(0, Math.min(gBottom, mBottom) - Math.max(g.y, m.y))

  return overlapX >= 50 && overlapY >= 50
}

// Parse persisted geometry JSON; returns null if missing/malformed. Clamps the
// size up to the window minimums.
export function parseGeometry(raw: string | null): Geometry | null {
  if (!raw) return null
  let g: unknown
  try {
    g = JSON.parse(raw)
  } catch {
    return null
  }
  if (typeof g !== 'object' || g === null) return null
  const rec = g as Record<string, unknown>
  if (
    typeof rec.x !== 'number' ||
    typeof rec.y !== 'number' ||
    typeof rec.width !== 'number' ||
    typeof rec.height !== 'number'
  ) {
    return null
  }
  return {
    x: rec.x,
    y: rec.y,
    width: Math.max(rec.width, MIN_WIDTH),
    height: Math.max(rec.height, MIN_HEIGHT)
  }
}

File: src/lazynote/geometry.py
"""Pure window-geometry helpers. Ported from the original geometry.ts.

No PySide6 imports — unit-testable under plain pytest.
"""

from __future__ import annotations

import json
from dataclasses import dataclass

MIN_WIDTH = 360
MIN_HEIGHT = 360


@dataclass
class Geometry:
    x: int
    y: int
    width: int
    height: int


def is_on_screen(g: Geometry, m: Geometry) -> bool:
    """True if the window overlaps the monitor by >= 50px on both axes."""
    overlap_x = max(0, min(g.x + g.width, m.x + m.width) - max(g.x, m.x))
    overlap_y = max(0, min(g.y + g.height, m.y + m.height) - max(g.y, m.y))
    return overlap_x >= 50 and overlap_y >= 50


def parse_geometry(raw: str | None) -> Geometry | None:
    """Parse persisted geometry JSON; None if missing/malformed. Clamps to minimums."""
    if not raw:
        return None
    try:
        data = json.loads(raw)
    except (ValueError, TypeError):
        return None
    if not isinstance(data, dict):
        return None
    if not all(isinstance(data.get(k), (int, float)) for k in ("x", "y", "width", "height")):
        return None
    return Geometry(
        x=int(data["x"]),
        y=int(data["y"]),
        width=max(int(data["width"]), MIN_WIDTH),
        height=max(int(data["height"]), MIN_HEIGHT),
    )
